matminer/featurizers/composition/roost.py

```diff
@@ -0,0 +1,569 @@
+        momentum=0.9,
+        weight_decay=1e-6,
+        batch_size=128,
+        workers=0,  # load data in main process -- allows caching
+        device=default_device,
+        **kwargs,
+    ):
+
+        # kill the featurizer if library not installed
+        if Roost == object:
+            raise ImportError(
+                "Roost library unavailable - please install and try again"
+            )
+
+        if task not in ["regression", "classification"]:
+            raise ValueError(
+                "Only 'regression' or 'classification' allowed for 'task'"
+            )
+
+        if n_targets is None:
+            if task == "regression":
+                n_targets = 1  # hard coded for single target regression
+            elif task == "classification":
+                n_targets = 2  # hard coded for binary clf
+
+        if elem_emb == "matscholar":
+            elem_emb_len = 200  # hard coded for matscholar
+        else:
+            raise NotImplementedError(
+                "Currently only the matscholar embedding is implemented"
+            )
+
+        self.data_params = {
+            "batch_size": batch_size,
+            "num_workers": workers,
+            "pin_memory": False,
+            "shuffle": True,
+            "collate_fn": collate_batch,
+        }
+
+        self.setup_params = {
+            "loss": loss,
+            "optim": optim,
+            "learning_rate": learning_rate,
+            "weight_decay": weight_decay,
+            "momentum": momentum,
+            "device": device,
+        }
+
+        self.model_params = {
+            "task": task,
+            "robust": False,
+            "n_targets": n_targets,
+            "elem_emb_len": elem_emb_len,
+            "elem_fea_len": elem_fea_len,
+            "n_graph": n_graph,
+            "elem_heads": elem_heads,
+            "elem_gate": elem_gate,
+            "elem_msg": elem_msg,
+            "cry_heads": cry_heads,
+            "cry_gate": cry_gate,
+            "cry_msg": cry_msg,
+            "out_hidden": out_hidden,
+        }
+
+        model, criterion, optimizer, scheduler, normalizer = init_roost(
+            **self.setup_params,
+            **self.model_params,
+        )
+
+        self.model_name = model_name
+        self.run_id = run_id
+
+        self.model = model
+        self.criterion = criterion
+        self.optimizer = optimizer
+        self.scheduler = scheduler
+        self.normalizer = normalizer
+
+        self.model_name = model_name
+        self.run_id = run_id
+
+        self.elem_fea_len = elem_fea_len
+        self.elem_emb = elem_emb
+        self.task = task
+        self.epochs = epochs
+
+        self.fitted = False
+
+        # NOTE multiprocessing doesn't play nicely with pytorch, apparently you
+        # can set the multiprocessing start method to spawn but it's safer just
+        # to only use one job?
+        # set_start_method('spawn', force=True)
+        self.set_n_jobs(1)
+
```

Comment:
Are there any compositions which this featurizer won't work for, and which can be checked easily?

E.g., if there is a composition with 15 elements, will roost still work?

If so, RoostFeaturizer needs a `precheck` method. See the `BaseFeaturizer.precheck` method for more details (reproduced here):

```
"""
        Precheck (provide an estimate of whether a featurizer will work or not)
        for a single entry (e.g., a single composition). If the entry fails the
        precheck, it will most likely fail featurization; if it passes, it is
        likely (but not guaranteed) to featurize correctly.

        Prechecks should be:
            * accurate (but can be good estimates rather than ground truth)
            * fast to evaluate
            * unlikely to be obsolete via changes in the featurizer in the near
                future

        This method should be overridden by any featurizer requiring its
        use, as by default all entries will pass prechecking. Also, precheck
        is a good opportunity to throw warnings about long runtimes (e.g., doing
        nearest neighbors computations on a structure with many thousand sites).

        See the documentation for precheck_dataframe for more information.

        Args:
            *x (Composition, Structure, etc.): Input to-be-featurized. Can be
                a single input or multiple inputs.

        Returns:
            (bool): True, if passes the precheck. False, if fails.

```

Reply:
It cannot deal with pure systems and it can't deal with compositions that include elements not present in the initial embedding dictionary.

It can scale up for something with 15 elements but the requirements do grow as O(n^2) within each material.